### PR TITLE
Remove Option to make API simpler and cleaner

### PR DESCRIPTION
**Pull request by slavikm**
## Description
Remove all the `WithOptions` from API and implementation

## Must
- [ ] Tests
- [ ] Documentation (if applicable)

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,6 +1,6 @@
 module github.com/descope/go-sdk
 
-go 1.18
```
Comment (orius123) [inline go.mod:3]: I think for SDKs we should support as far back as we can

**Review by orius123 — COMMENTED** [review]
I don't understand the motivation for this PR, it will make it harder for us to extend and add new options without breaking the function signatures.
And will require customers to pass Nils all around

**Comment by aviadl** [comment]
> I don't understand the motivation for this PR, it will make it harder for us to extend and add new options without breaking the function signatures.
> And will require customers to pass Nils all around 

The implementation of options was not really an option, and usually when you do this (a trick to path an arbitrary number of arguments) it is meant for configuration 
So we thought it would be better to remove it all together

@slavikm maybe we can take an approach that creates a param object that has the relevant function call on it
Then you can extend the signature without breaking rather easily 

But don't recall seeing that notion on other libs

File: go.mod (hunk 1 above)
Comment (slavikm) [reply go.mod:3]: You might be right here. I kind of jumped the gun because I saw renovate trying and failing.

**Comment by slavikm** [comment]
> I don't understand the motivation for this PR, it will make it harder for us to extend and add new options without breaking the function signatures. And will require customers to pass Nils all around

The current implementation is the worst of both worlds. Does not allow extension of the code and is very brittle with `interface{}` and casting. I rather make it simpler and if needed add another parameter to the relevant function than dirty the entire code with weirdness.

File: descope/auth/auth.go [hunk 2]
```diff
@@ -458,7 +448,7 @@ func (*authenticationsBase) verifyDeliveryMethod(method DeliveryMethod, identifi
 	return nil
 }
 
-func (auth *authenticationsBase) exchangeTokenWithOptions(code string, url string, options ...Option) (*AuthenticationInfo, error) {
+func (auth *authenticationsBase) exchangeTokenWithOptions(code string, url string, w http.ResponseWriter) (*AuthenticationInfo, error) {
```
Comment (aviadl) [inline descope/auth/auth.go:451]: we should rename this function name
We have no options here anymore :)

File: descope/auth/sdk.go [hunk 3]
```diff
@@ -42,18 +46,11 @@ type MagicLink interface {
 	SignUpOrInCrossDevice(method DeliveryMethod, identifier string, URI string) (*MagicLinkResponse, error)
 
 	// GetSession - Use to get a session that was generated by SignIn/SignUp request, and verified with Verify request.
-	// This is a shortcut for GetSessionWithOptions(method, code, WithResponseOption(w))
 	GetSession(pendingRef string, w http.ResponseWriter) (*AuthenticationInfo, error)
-	// GetSessionWithOptions - Use to get a session that was generated by SignInMagicLink/SignUpMagicLink request, and verified with VerifyMagicLink request.
```
Comment (aviadl) [inline descope/auth/sdk.go:47]: Need to add this comment back

File: descope/auth/sdk.go [hunk 4]
```diff
@@ -42,18 +46,11 @@ type MagicLink interface {
...
-	GetSessionWithOptions(pendingRef string, options ...Option) (*AuthenticationInfo, error)
 
 	// Verify - Use to verify a SignIn/SignUp request, based on the magic link token generated.
 	// if the link was generated with crossDevice, the authentication info will be nil, and should returned with GetSession.
-	// This is a shortcut for VerifyWithOptions(method, code, WithResponseOption(w))
 	Verify(token string, w http.ResponseWriter) (*AuthenticationInfo, error)
-	// VerifyWithOptions - used to verify a SignIn/SignUp request, based on the magic link token generated.
```
Comment (aviadl) [inline descope/auth/sdk.go:54]: and this one

File: descope/auth/sdk.go [hunk 5]
```diff
@@ -96,12 +96,7 @@ type OTP interface {
 	// In case the request cookie can be renewed an automatic renewal is called and returns a new set of cookies to use.
 	// Use the ResponseWriter (optional) to apply the cookies to the response automatically.
 	// returns a list of cookies or an error upon failure.
-	// This is a shortcut for VerifyCodeWithOptions(method, identifier, code, WithResponseOption(w))
 	VerifyCode(method DeliveryMethod, identifier string, code string, w http.ResponseWriter) (*AuthenticationInfo, error)
-	// VerifyCodeWithOptions - used to verify a SignIn/SignUp based on the given identifier either an email or a phone
```
Comment (aviadl) [inline descope/auth/sdk.go:101]: ...

File: descope/auth/sdk.go [hunk 6]
```diff
@@ -138,35 +132,22 @@ type OAuth interface {
 	// returns an error upon failure and a string represent the redirect URL upon success.
 	// Uses the response writer to automatically redirect the client to the provider url for authentication.
 	// A successful authentication will result in a callback to the url defined in the current project settings.
-	// This is a shortcut for StartWithOptions(provider, WithResponseOption(w))
 	Start(provider OAuthProvider, returnURL string, w http.ResponseWriter) (string, error)
-	// OAuthStartWithOptions - use to start an OAuth authentication using the given OAuthProvider and options.
```
Comment (aviadl) [inline descope/auth/sdk.go:143]: ...

File: descope/auth/sdk.go [hunk 7]
```diff
@@ -175,35 +156,30 @@ type WebAuthn interface {
 	// by calling document.location.origin via javascript.
 	// returns a transaction id response on success and error upon failure.
 	SignUpStart(identifier string, user *User, origin string) (*WebAuthnTransactionResponse, error)
+
 	// SignUpFinish - Use to finish an authentication process with a given transaction id and credentials after been signed
 	// by the credentials navigator.
 	// Use the ResponseWriter (optional) to apply the cookies to the response automatically.
-	// This is a shortcut for SignUpFinishWithOptions(finishRequest, WithResponseOption(w))
 	SignUpFinish(finishRequest *WebAuthnFinishRequest, w http.ResponseWriter) (*AuthenticationInfo, error)
-	// SignUpWebAuthnFinishWithOptions - Use to finish an authentication process with a given transaction id and credentials after been signed
```
Comment (aviadl) [inline descope/auth/sdk.go:183]: ...

File: descope/auth/sdk.go [hunk 8]
```diff
@@ -175,35 +156,30 @@ type WebAuthn interface {
...
 	// by calling document.location.origin via javascript.
 	// returns a transaction id response on successs and error upon failure.
 	SignInStart(identifier string, origin string) (*WebAuthnTransactionResponse, error)
+
 	// SignInFinish - Use to finish an authentication process with a given transaction id and credentials after been signed
 	// by the credentials navigator.
 	// Use the ResponseWriter (optional) to apply the cookies to the response automatically.
-	// This is a shortcut for SignInFinishWithOptions(finishRequest, WithResponseOption(w))
 	SignInFinish(finishRequest *WebAuthnFinishRequest, w http.ResponseWriter) (*AuthenticationInfo, error)
-	// SignInWebAuthnFinishWithOptions - Use to finish an authentication process with a given transaction id and credentials after been signed
```
Comment (aviadl) [inline descope/auth/sdk.go:197]: ...

File: descope/auth/sdk.go (hunk 3 above)
Comment (slavikm) [reply descope/auth/sdk.go:47]: Why? I have the same exact comment above

File: descope/auth/sdk.go (hunk 5 above)
Comment (slavikm) [reply descope/auth/sdk.go:101]: Why? I have the same exact comment above

**Review by aviadl — APPROVED** [review]

**Comment by aviadl** [comment]
@slavikm merging this, so we can make progress
But I still think that we lost some of the SDK descriptions, so suggest that you review them again

File: descope/auth/sdk.go [hunk 9]
```diff
@@ -248,25 +220,19 @@ type Authentication interface {
...
 
 	// DeleteCookies - Deletes the session and refresh cookies in the http response.
 	// Use the ResponseWriter (optional) to apply the cookies to the response automatically.
-	// This is a shortcut for DeleteCookiesWithOptions(r, WithResponseOption(w))
 	DeleteCookies(request *http.Request, w http.ResponseWriter) error
-	// DeleteCookiesWithOptions - Deletes the session and refresh cookies in the http response.
-	DeleteCookiesWithOptions(request *http.Request, options ...Option) error
 
 	// Logout - Use to perform logout from all active sessions for the request user. This will revoke the given tokens
 	// and if given options will also remove existing session on the given response sent to the client.
```
Comment (Bars92) [inline descope/auth/sdk.go:233]: we may need to go over the docs again and make sure no options leftovers are still documented @slavikm @aviadl

**Pull request merged**